wallet/db_postgres.c

```diff
@@ -1,34 +1,264 @@
+}
+
+static bool db_postgres_column_is_null(struct db_stmt *stmt, int col)
+{
+	PGresult *res = (PGresult*)stmt->inner_stmt;
+	return PQgetisnull(res, stmt->row, col);
+}
+
+static u64 db_postgres_column_u64(struct db_stmt *stmt, int col)
+{
+	PGresult *res = (PGresult*)stmt->inner_stmt;
+	be64 bin;
+	size_t expected = sizeof(bin), actual = PQgetlength(res, stmt->row, col);
+
+	if (expected != actual)
+		db_fatal(
+		    "u64 field doesn't match size: expected %zu, actual %zu\n",
+		    expected, actual);
+
+	memcpy(&bin, PQgetvalue(res, stmt->row, col), sizeof(bin));
+	return be64_to_cpu(bin);
+}
+
+static s64 db_postgres_column_int(struct db_stmt *stmt, int col)
+{
+	PGresult *res = (PGresult*)stmt->inner_stmt;
+	be32 bin;
+	size_t expected = sizeof(bin), actual = PQgetlength(res, stmt->row, col);
+
+	if (expected != actual)
+		db_fatal(
+		    "s32 field doesn't match size: expected %zu, actual %zu\n",
+		    expected, actual);
+
+	memcpy(&bin, PQgetvalue(res, stmt->row, col), sizeof(bin));
+	return be32_to_cpu(bin);
+}
+
+static size_t db_postgres_column_bytes(struct db_stmt *stmt, int col)
+{
+	PGresult *res = (PGresult *)stmt->inner_stmt;
+	return PQgetlength(res, stmt->row, col);
+}
+
+static const void *db_postgres_column_blob(struct db_stmt *stmt, int col)
+{
+	PGresult *res = (PGresult*)stmt->inner_stmt;
+	return PQgetvalue(res, stmt->row, col);
+}
+
+static const unsigned char *db_postgres_column_text(struct db_stmt *stmt, int col)
+{
+	PGresult *res = (PGresult*)stmt->inner_stmt;
+	return (unsigned char*)PQgetvalue(res, stmt->row, col);
+}
+
+static void db_postgres_stmt_free(struct db_stmt *stmt)
+{
+	if (stmt->inner_stmt)
+		PQclear(stmt->inner_stmt);
+	stmt->inner_stmt = NULL;
+}
+
+static bool db_postgres_exec(struct db_stmt *stmt)
+{
+	bool ok;
+	stmt->inner_stmt = db_postgres_do_exec(stmt);
+	ok = PQresultStatus(stmt->inner_stmt) == PGRES_COMMAND_OK;
+
+	if (!ok)
+		stmt->error = PQerrorMessage(stmt->db->conn);
+
+	return ok;
+}
+
+static u64 db_postgres_last_insert_id(struct db_stmt *stmt)
+{
+	PGresult *res = PQexec(stmt->db->conn, "SELECT lastval()");
+	int id = atoi(PQgetvalue(res, 0, 0));
+	PQclear(res);
+	return id;
+}
+
+static size_t db_postgres_count_changes(struct db_stmt *stmt)
+{
+	PGresult *res = (PGresult*)stmt->inner_stmt;
+	char *count = PQcmdTuples(res);
+	return atoi(count);
+}
+
+static void db_postgres_teardown(struct db *db)
+{
+}
 
 struct db_config db_postgres_config = {
-	.name = "postgres",
-	.queries = NULL,
-	.num_queries = DB_POSTGRES_QUERY_COUNT,
-	.exec_fn = NULL,
-	.query_fn = NULL,
-	.step_fn = NULL,
-	.begin_tx_fn = NULL,
-	.commit_tx_fn = NULL,
-	.stmt_free_fn = NULL,
-
-	.column_is_null_fn = NULL,
-	.column_u64_fn = NULL,
-	.column_int_fn = NULL,
-	.column_bytes_fn = NULL,
-	.column_blob_fn = NULL,
-	.column_text_fn = NULL,
-
-	.last_insert_id_fn = NULL,
-	.count_changes_fn = NULL,
-	.setup_fn = NULL,
-	.teardown_fn = NULL,
+    .name = "postgres",
+    .queries = db_postgres_queries,
```

Comment:
tabs to spaces gratuitous change?

Reply:
Hm, seems a strange switch in my clang-format configuration. I don't remember, what's the canonical indentation mechanism for c-lightning? Not to get into a tabs-vs-spaces war here :stuck_out_tongue_winking_eye:

Reply:
One tab character. Indents to align text on previous lines use a mix of tabs followed by 0 to 7 spaces.

Reply:
Ok, will need to adjust my `.clang-format` file accordingly.